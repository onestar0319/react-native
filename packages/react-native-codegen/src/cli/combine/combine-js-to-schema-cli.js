/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+react_native
 * @flow strict-local
 * @format
 */

'use strict';

const combine = require('./combine-js-to-schema');
const fs = require('fs');
// $FlowFixMe[untyped-import] glob is untyped
const glob = require('glob');
const path = require('path');

const [outfile, ...fileList] = process.argv.slice(2);

const allFiles = [];
fileList.forEach(file => {
  if (fs.lstatSync(file).isDirectory()) {
    const dirFiles = glob
      .sync(`${file}/**/*.js`, {
        nodir: true,
      })
      .filter(
        f =>
          /^(Native.+|.+NativeComponent)/.test(path.basename(f)) &&
          // NativeUIManager will be deprecated by Fabric UIManager.
          // For now, ignore this spec completely because the types are not fully supported.
          !f.endsWith('NativeUIManager.js') &&
          !f.includes('__tests'),
      );
    allFiles.push(...dirFiles);
  } else {
    allFiles.push(file);
  }
});

const formattedSchema = JSON.stringify(combine(allFiles), null, 2);
if (outfile != null) {
  fs.writeFileSync(outfile, formattedSchema);
} else {
  console.log(formattedSchema);
}
