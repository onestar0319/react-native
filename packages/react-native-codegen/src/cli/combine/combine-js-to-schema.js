/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

'use strict';
import type {SchemaType} from '../../CodegenSchema.js';

const SchemaParser = require('../../parsers/schema');

function combineSchemas(files: Array<string>): SchemaType {
  return files.reduce(
    (merged, filename) => {
      const schema = SchemaParser.parse(filename);
      if (schema && schema.modules) {
        merged.modules = {...merged.modules, ...schema.modules};
      }
      return merged;
    },
    {modules: {}},
  );
}

module.exports = combineSchemas;
