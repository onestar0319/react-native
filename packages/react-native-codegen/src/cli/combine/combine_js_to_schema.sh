#!/bin/bash
# Copyright (c) Meta Platforms, Inc. and affiliates.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.

set -e
set -u

THIS_DIR=$(cd -P "$(dirname "$(realpath "${BASH_SOURCE[0]}" || echo "${BASH_SOURCE[0]}")")" && pwd)

SETUP_ENV_VARS_PATH="$BUCK_PROJECT_ROOT/env-utils/setup_env_vars.sh"

# Somehow this path this doesn't work in BUCK1, so falling back to the exisiting path that we use in BUCK1.
if [ ! -f "${SETUP_ENV_VARS_PATH}" ]; then
  SETUP_ENV_VARS_PATH="$BUCK_PROJECT_ROOT/xplat/js/env-utils/setup_env_vars.sh"
fi

# shellcheck source=xplat/js/env-utils/setup_env_vars.sh
source "${SETUP_ENV_VARS_PATH}"

pushd "$JS_DIR" >/dev/null
  "$INSTALL_NODE_MODULES"
popd >/dev/null

exec "$FLOW_NODE_BINARY" "$THIS_DIR/combine-js-to-schema-cli.js" "$@"
