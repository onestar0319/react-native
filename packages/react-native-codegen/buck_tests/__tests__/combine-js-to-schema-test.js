/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @emails oncall+react_native
 * @flow strict-local
 * @format
 */

import combine from '../combine-js-to-schema';

jest.mock(
  '/test/module/SchemaOne',
  () => ({
    modules: {
      ComponentOne: {
        foo: 'baz',
      },
    },
  }),
  {virtual: true},
);

jest.mock(
  '/test/module/SchemaTwo',
  () => ({
    modules: {
      ComponentTwo: {
        foo: 'bar',
      },
    },
  }),
  {virtual: true},
);

test('should combine files', () => {
  const files = ['/test/module/SchemaOne', '/test/module/SchemaTwo'];
  expect(combine(files)).toMatchSnapshot();
});

test('should throw for failed require', () => {
  const files = ['/test/module/does/not/exist'];
  expect(() => combine(files)).toThrow(
    "Can't require file at /test/module/does/not/exist",
  );
});
