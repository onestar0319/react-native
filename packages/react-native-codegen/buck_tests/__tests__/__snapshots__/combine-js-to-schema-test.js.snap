// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should combine files 1`] = `
Object {
  "modules": Object {
    "ComponentOne": Object {
      "foo": "baz",
    },
    "ComponentTwo": Object {
      "foo": "bar",
    },
  },
}
`;
